### PR TITLE
Embed protocols back in Kuzzle & refactor functional tests

### PR DESCRIPTION
**Pull request by benoitvidis**
This PR enable Kuzzle for working in stand-alone mode, without any third-party proxy.
Protocols are now directly integrated inside Kuzzle.

### config & backward compatibility

If any configuration related to previous proxy mode (`proxyBroker` service), Kuzzle will disable the embedded mode and activate the proxy mode only.

### timeouts

When using an external Kuzzle proxy, clients can recieve timeout errors if the proxy considers Kuzzle si tool long to respond.
After discussion, this feature has been removed from the internal entry point.

### compose stacks

Both dev & test compose stacks expose both entrypoints (7512 is bound to Kuzzle, 7513 to the proxy).

## Functional tests refactor

Now that Kuzzle needs to support 2 entry points, functional tests needed to be adapted to test both of them.
We jumped on the opportunity to add a *real* websocket protocol client and to upgrade `cucumberjs` to the 2.x branch.

* most of the code migrated to ES2015 syntax
* Feature files are shared across all protocols & endpoints. The connector & enpoints to use are set based on the options passed to the cli
* 6 profiles are now up (=cli arguments): 
  * httpEmbedded
  * websocketEmbedded
  * socketioEmbedded
  * httpProxy
  * websocketProxy
  * socketioProxy
* `protocol`, `host` and `port` parameters can be passed via [`world-parameters`](https://github.com/cucumber/cucumber-js/blob/master/docs/cli.md#world-parameters) (cf example below)

```
./node_modules/.bin/cucumberjs -p websocketEmbedded

./node_modules/.bin/cucumberjs -p socketioProxy --world-parameters '{"host": "172.16.0.4", "port": 7777}'
```

## Related changes

* [x] [mqtt protocol](https://github.com/kuzzleio/protocol-mqtt) plugin needs an embed version (current version only works with proxy)
* [ ] documentation to come
* [x] [compose stacks need to be updated to not use the proxy by default](https://github.com/kuzzleio/kuzzle-build/pull/14)

### CODE REVIEW

File: lib/api/core/entrypoints/index.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const
+  Bluebird = require('bluebird'),
+  Embedded = require('./embedded'),
+  Proxy = require('./proxy');
```
Comment (kuzzle) [inline lib/api/core/entrypoints/index.js]: MAJOR Remove this override of "Proxy". [rule](http://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS2424)

File: lib/api/core/entrypoints/embedded/index.js [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+      connection = this.clients[request.context.connectionId];
+
+    if (!connection) {
+      return this.kuzzle.pluginsManager.trigger('log:warn', `[access log] No connection retrieved for connection id: ${request.context.connectionId}\n` +
+        'Most likely, the connection was closed before the response was received.');
+    }
+
+    if (this.config.logs.accessLogFormat === 'logstash') {
+      // custom kuzzle logs to be exported to logstash
+      this.logger.info({
```
Comment (kuzzle) [inline lib/api/core/entrypoints/embedded/index.js]: MINOR Group all shorthand properties at the beginning of this object declaration [rule](http://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3499)

File: lib/api/core/entrypoints/embedded/index.js [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+    }
+
+    const protocol = connection.protocol.toUpperCase();
+
+    // = apache combined
+    let
+      verb = 'DO',
+      url = '',
+      user,
+      status = request.status;
```
Comment (kuzzle) [inline lib/api/core/entrypoints/embedded/index.js]: CRITICAL Make "status" "const". [rule](http://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3353)

**Comment by kuzzle** [comment]
SonarQube analysis reported 3 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[CRITICAL]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Comment by benoitvidis** [comment]
sonarcube + protocols init should return a promise => wip

File: test/api/core/pluginContext/pluginContext.test.js [hunk 4]
```diff
@@ -15,9 +15,20 @@ describe('Plugin Context', () => {
...
 
   beforeEach(() => {
+    mockrequire('../../../../lib/services/internalEngine', function () {
+      this.init = sinon.spy();
+      this.bootstrap = {
+        all: sinon.spy(),
+        createCollection: sinon.spy()
+      };
+    });
+    mockrequire.reRequire('../../../../lib/api/core/plugins/pluginContext');
```
Comment (benoitvidis) [inline test/api/core/pluginContext/pluginContext.test.js]: Boyscouting - InternalEngine was trying to init a connection to ES

**Comment by benoitvidis** [comment]
protocol plugins need some additional context => wip

File: lib/api/core/entrypoints/embedded/context.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const
+  ClientConnection = require('./clientConnection'),
+  debug = require('../../../../kuzzleDebug')('kuzzle:entry-point:protocols'),
+  errors = require('kuzzle-common-objects').errors,
```
Comment (kuzzle) [inline lib/api/core/entrypoints/embedded/context.js]: MINOR Use destructuring syntax for these assignments from "require('kuzzle-common-objects')". [rule](http://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3514)

File: lib/api/core/entrypoints/embedded/index.js [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+   * @param {string} connectionId
+   */
+  removeConnection (connectionId) {
+    if (!this.clients[connectionId]) {
+      return;
+    }
+
+    const connection = this.clients[connectionId];
+    this.kuzzle.router.removeConnection(new RequestContext({
+      connectionId: connectionId,
```
Comment (kuzzle) [inline lib/api/core/entrypoints/embedded/index.js]: MINOR Use shorthand for property "connectionId". [rule](http://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS3498)

**Comment by kuzzle** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: .kuzzlerc.sample [hunk 7]
```diff
@@ -212,7 +212,111 @@
...
+      "embedded": true,
+      "proxy": false
+    },
+
+    // [proxy]
+    // A Kuzzle proxy can be used to forward requests to Kuzzle.
+    //   * host:
+    //       Host name/address of a Kuzzle proxy server.
+    //       Can be a hostname, an IP address or an URI.
+    //   * pingTimeout:
```
Comment (ballinette) [inline .kuzzlerc.sample]: nitpicking: should say the default value

File: .kuzzlerc.sample [hunk 8]
```diff
@@ -212,7 +212,111 @@
...
+    //   * port:
+    //     The port on which Kuzzle should bind to
+    //
+    "maxRequestSize": "1mb",
+    "port": 7512,
+    // [protocols] (embedded only)
+    // protocols accepted by Kuzzle.
+    // protocols can be extended using plugins and configured in this section.
+    "protocols": {
+      // * maxFormFileSize:
```
Comment (ballinette) [inline .kuzzlerc.sample]: nitpicking: move the comment just above the related entry would be more readable:
```
"protocols": {
  "http": {
    "enable": true,
    // maximum size of documents sent via http forms:
    "maxFormFileSize": "1mb"
  },
  "socketio": {
    "enable": true,
    // value of Access-Control-Allow-Origin header to answer the upgrade request:
    "origins": '*:*' 
  },
  "websocket": {
    "enable": true
  }
}
```

File: docker-compose/dev.yml [hunk 9]
```diff
@@ -2,9 +2,9 @@ version: '3'
 
 services:
   proxy:
-    image: kuzzleio/proxy:develop
+    image: kuzzleio/proxy
     ports:
-      - "7512:7512"
+      - "7513:7512"
```
Comment (ballinette) [inline docker-compose/dev.yml]: see env variables below:
```
kuzzle_server__entryPoints__proxy=true
kuzzle_server__proxy__host=proxy 
```

File: lib/api/core/entrypoints/embedded/index.js [hunk 10]
```diff
@@ -0,0 +1,464 @@
...
+  RequestContext = require('kuzzle-common-objects').models.RequestContext,
+  winston = require('winston'),
+  WinstonElasticsearch = require('winston-elasticsearch'),
+  WinstonSyslog = require('winston-syslog');
+
+class EmbeddedEntryPoint extends EntryPoint {
+  constructor (kuzzle) {
+    super(kuzzle);
+
+    this.kuzzle = kuzzle;
```
Comment (ballinette) [inline lib/api/core/entrypoints/embedded/index.js]: Nitpicking: `this.kuzzle` is already set in superclass constructor, no need to set here again.

File: .kuzzlerc.sample (hunk 7 above)
Comment (benoitvidis) [inline .kuzzlerc.sample]: default values are the ones set in the file.

File: lib/api/core/entrypoints/embedded/index.js [hunk 11]
```diff
@@ -0,0 +1,464 @@
...
+      this.protocols.websocket = new WebSocketProtocol();
+      this.protocols.websocket.init(this);
+    }
+
+    if (this.config.protocols.socketio.enabled) {
+      this.protocols.socketio = new SocketIoProtocol();
+      this.protocols.socketio.init(this);
+    }
+
+    return Bluebird.resolve()
```
Comment (ballinette) [inline lib/api/core/entrypoints/embedded/index.js]: why not simply 
```javascript
return this.loadMoreProtocols()
  .catch(e => {
    this.kuzzle.pluginsManager.trigger('log:error', e);
    throw e;
  });
```

File: .kuzzlerc.sample (hunk 7 above)
Comment (ballinette) [inline .kuzzlerc.sample]: but `pingTimeout` and `pingInterval` are not present in the sample file

File: lib/api/core/entrypoints/embedded/protocols/websocket.js [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+ * @class WebsocketProtocol
+ */
+class WebsocketProtocol extends Protocol {
+  constructor () {
+    super();
+
+    this.channels = {};
+    this.connectionPool = {};
+    this.server = null;
+    this.protocol = 'websocket';
```
Comment (ballinette) [inline lib/api/core/entrypoints/embedded/protocols/websocket.js]: nitpicking: here we have a class property `this.protocol`, while in [socketio](https://github.com/kuzzleio/kuzzle/blob/no-proxy/lib/api/core/entrypoints/embedded/protocols/socketio.js#L25), we use a module variable `_protocol`.
This should be harmonized.

File: lib/api/core/httpRouter/index.js [hunk 13]
```diff
@@ -143,33 +143,33 @@ class Router {
...
+    if (httpRequest.content.length <= 0) {
+      return routeHandler.invokeHandler(cb);
+    }
+
+    if (httpRequest.headers['content-type']
+      && !httpRequest.headers['content-type'].startsWith('application/json')) {
+      return replyWithError(cb, routeHandler.getRequest(), new BadRequestError(`Invalid request content-type. Expected "application/json", got: "${httpRequest.headers['content-type']}"`));
+    }
+
+    {
```
Comment (ballinette) [inline lib/api/core/httpRouter/index.js]: why this brackets ?

File: docker-compose/dev.yml (hunk 9 above)
Comment (stafyniaksacha) [inline docker-compose/dev.yml]: :+1:

File: .kuzzlerc.sample (hunk 7 above)
Comment (benoitvidis) [inline .kuzzlerc.sample]: ah true. Sorry, I had missed it

**Comment by benoitvidis** [comment]
Checking an issue with ip offset & access log for http > wip

**Pull request merged**